File: engine/common/core/src/cpp/Log.cpp
#include <iostream>
#include "Log.h"

namespace ae {
    
/** */
Logger *Log::logger = (Logger *)0;
    
/**  */
void Log::trace(const char *tag,const char *msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->trace(tag,msg);
}

/**  */
void Log::trace(const char *tag,const std::string& msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->trace(tag,msg.c_str());
}
    
/**  */
void Log::debug(const char *tag,const char *msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->debug(tag,msg);
}
    
/**  */
void Log::debug(const char *tag,const std::string& msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->debug(tag,msg.c_str());
}

/**  */
void Log::info(const char *tag,const char *msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->info(tag,msg);
}
    
/**  */
void Log::info(const char *tag,const std::string& msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->info(tag,msg.c_str());
}
    
/**  */
void Log::warning(const char *tag,const char *msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->warning(tag,msg);
}
    
/**  */
void Log::warning(const char *tag,const std::string& msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->warning(tag,msg.c_str());
}
    
/**  */
void Log::error(const char *tag,const char *msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->error(tag,msg);
}
    
/**  */
void Log::error(const char *tag,const std::string& msg) {
    if (logger == (Logger *)0) {
        return;
    }
    logger->error(tag,msg.c_str());
}
    
} // namespace

File: engine/common/core/src/lua/ae/log.lua
-- @brief Logs message format using `string.format`.

-- null logger if none set
if not log then
    log = {
        trace = function()
        end,
        debug = function()
        end,
        info = function()
        end,
        warning = function()
        end,
        error = function()
        end
    }
end

-- keep the native library
local _log = log

-- the new library
local log = {}

-- @func
-- @brief Logs a trace message.
-- @param message The message.
-- @func
-- @brief Logs a trace message.
-- @param message The message.
-- @param args The list of format arguments.
function log.trace(...)
    _log.trace(string.format(...))
end

-- @func
-- @brief Logs a debug message.
-- @param message The message.
-- @func
-- @brief Logs a debug message.
-- @param message The message.
-- @param args The list of format arguments.
function log.debug(...)
    _log.debug(string.format(...))
end

-- @func
-- @brief Logs an info message.
-- @param message The message.
-- @func
-- @brief Logs an info message.
-- @param message The message.
-- @param args The list of format arguments.
function log.info(...)
    _log.info(string.format(...))
end

-- @func
-- @brief Logs a warning message.
-- @param message The message.
-- @func
-- @brief Logs a warning message.
-- @param message The message.
-- @param args The list of format arguments.
function log.warning(...)
    _log.warning(string.format(...))
end

-- @func
-- @brief Logs an error message.
-- @param message The message.
-- @func
-- @brief Logs an error message.
-- @param message The message.
-- @param args The list of format arguments.
function log.error(...)
    _log.error(string.format(...))
end

return log